rbd.c

```diff
@@ -215,39 +252,39 @@ static int timer_check_and_set_def(struct tcmu_device *dev)
 	tcmu_dev_dbg(dev, "The cluster's default osd op timeout(%f), osd heartbeat grace(%d) interval(%d)\n",
 		     timeout, grace, interval);
 
+	cmd_time_out = tcmu_get_dev_cmd_time_out(dev);
+
 	/* Frist: Try to use new osd op timeout value */
-	if (state->osd_op_timeout && atof(state->osd_op_timeout) > grace + interval)
+	if (state->osd_op_timeout && atof(state->osd_op_timeout) > grace + interval) {
+		/* Make sure osd_op_timeout is larger than cmd_time_out */
+		if (atoi(state->osd_op_timeout) <= cmd_time_out) {
```

Comment:
If cmd_time_out is turned off (0) then this will increase the osd timeout when it shouldn't, so you need an extra "cmd_time_out > 0 && other test".

Are you guys setting cmd_time_out to catch tcmu-runner hangs/crashes? What is the initiator behavior when this happens? It seems like with multipathing you are going to get odd behavior where runner is down, tcmu will timeout the command and send the initiator a response. That will allow the initiator and multipath layer to complete its error handling if it fired and end up retrying this same path. Since runner is down, more IO will be sent and it will timeout and possibly kick off the eh again and we repeat over and over. You then have to restart runner and possibly the box because runner's restart code is buggy.

If runner is down due to crash or bug with cmd_time_out=0 then IO is stuck, it will time out on the initiator side and it will end up marking the path as bad. We will not have the ping ponging. You will still have to reboot, becuase the tcmu-runner/tcmu code is buggy for restart in this case.

When we fix the runner restart code, then with cmd_time_out=0 you will not have the ping ponging between paths.

Reply:
If cmd_time_out is turned off (0) then this will increase the osd timeout when it shouldn't, so
    you need an extra "cmd_time_out > 0 && other test".
Yes, this should be fixed.

Reply:
@mikechristie
I think I have made a mistake, the osd_op_timeout should be < cmd_time_out for this case.

Reply:
Our test case and environment are:
1, Ceph cluster with the size 30GB totailly and with creating 2 rbd devices(16G for each)
2, Only one tcmu node with two net interface: 10.XX.XX.222 and 126.XX.XX.222, will map two LUNs for each target devices.
3, The cmd_time_out is 3s, the osd op timeout is 30s.

Firstly write one rbd device with random data almost full, then try writing random data to another rbd device using multipath. The ceph cluster will be stuck when just the second target will be almost full, then remove the first rbd device from the cluster and the cluster will recovery.

There still has enough empty area in the 1G Data area of the ring buffer for the rest of IOs, so even the cluster is stuck the initiator could send all the rest of IOs cmds successfully.

if cmd_time_out < osd op timeout, when the old cmd has timed out in ring buffer and the initiator may retry the other path when it must. While at the same time the cluster recovers and the new one cmd comes, does there has problems in ceph cluster with two same IO operations at the same time ?

I think there also has the same problems with two different tcmu nodes use the same produce steps here.